Implement `absolute=False` URL validation

sirosen opened the pull request:
- Add a new argument `absolute`, which defaults to True. It can be set to False only when `relative` is also set to True, and allows validation of relative-only URLs.
- The regex generator used inside of the URL validator is expanded slightly to a more imperative style, making it easier to reason about the different branches it takes.
- New tests are added to confirm the new behavior, showing that parsing of relative and relative-only URLs works correctly and rejects valid absolute URLs when absolute=False

Resolves #2118

---

I looked at the issue tracker for an easy win and this one seemed pretty straightforward to me.
However, I only gave the [URL spec from WHATWG](https://url.spec.whatwg.org/) a pretty cursory glance, so it's very possible that there are missing cases.

sirosen committed:
Implement `absolute=False` URL validation

- Add a new argument `absolute`, which defaults to True. It can be set
  to False only when `relative` is also set to True, and allows
  validation of relative-only URLs.
- The regex generator used inside of the URL validator is expanded
  slightly to a more imperative style, making it easier to reason
  about the different branches it takes.
- New tests are added to confirm the new behavior, showing that
  parsing of relative and relative-only URLs works correctly and
  rejects valid absolute URLs when absolute=False

Resolves #2118

lafrech commented on src/marshmallow/validate.py:
```diff
@@ -149,11 +166,17 @@ def __init__(
+        if not relative and not absolute:
+            raise ValueError(
+                "URL validation cannot set both relative and absolute to False."
+            )
```
Should we also prevent `absolute=False` and `require_tld=True`?

sirosen replied:
Yeah, that would make sense. I'll add that shortly. I also realized that I'm missing a test case for this, so I'll be sure to test both of these errors.

sirosen replied:
I went to do this and realized that it won't work cleanly. `require_tld` defaults to `True`, so forbidding the combination makes `URL(absolute=False, relative=True)` fail.

I'd rather allow that usage than catch `absolute=False, require_tld=True` (explicit usage) with an error, so I'm going to leave this change out. But I'll add a test for the existing ValueError.

lafrech commented on src/marshmallow/validate.py:
```diff
@@ -105,38 +106,54 @@ class RegexMemoizer:
+                    # scheme
+                    r"(?:[a-z0-9\.\-\+]*)://",  # scheme is validated separately
+                    # basic_auth
+                    r"(?:[^:@]+?(:[^:@]*?)?@|)",  # basic auth
```
Tanks for adding comments. Maybe put all comments in their own line for consistency.

sirosen replied:
Sure, I'll do a quick cleanup pass.

lafrech reviewed (commented):
Thanks.

sirosen committed:
Improve URL.absolute and add it to repr

Three small changes:

Refactor the regex building slightly to better clarify the way that
the various hostname expressions are combined, improve the comments,
and remove unnecessary leading underscores from variable names.
(No functional change.)

Add a test for `absolute=False, relative=False` to confirm the
ValueError.

Add `absolute` to the URL validator repr.

sirosen commented:
I made a few small changes here to improve things.
- Better comments and a readability refactor in the regex builder.
- Add test for `relative=False, absolute=False`
- Add `absolute` to the `URL` repr

lafrech commented on src/marshmallow/validate.py:
```diff
@@ -109,40 +109,54 @@ def __init__(self):
             self, relative: bool, absolute: bool, require_tld: bool
         ) -> typing.Pattern:
-            _absolute_part = "".join(
+            hostname_variants = [
```
Nitpick: I suppose you could use a tuple here as well

sirosen replied:
I think the list is slightly better because there's a conditional append. If you feel strongly that it should be a tuple, I can make that change, but I'm inclined to keep it a list since it's an implementation detail.

lafrech reviewed (approved):
Thanks!

lafrech reviewed (approved):
LGTM.

Thanks!
